src/DashboardService.Data/ChangeLogRepository.cs

```diff
@@ -19,28 +21,102 @@ public ChangeLogRepository(IDataProvider provider)
+  public async Task<DbChangeLog> GetAsync(Guid id, CancellationToken ct)
   {
-    throw new NotImplementedException();
+    IQueryable<DbChangeLog> changeLog = _provider.ChangeLogs.AsNoTracking();
+
+    return await changeLog.FirstOrDefaultAsync(db => db.Id == id);
   }
 
-  public Task<List<DbChangeLog>> GetChangeLogsAsync(GetChangeLogsFilter filter, CancellationToken ct)
+  public async Task<(List<DbChangeLog>, int totalCount)> GetChangeLogsAsync(GetChangeLogsFilter filter, CancellationToken ct)
   {
-    throw new NotImplementedException();
+    IQueryable<DbChangeLog> changeLogs = _provider.ChangeLogs.AsNoTracking();
+
+    if (filter.TaskId.HasValue)
+    {
+      changeLogs = changeLogs.Where(d => d.TaskId == filter.TaskId);
+    }
+
+    if (filter.IsAscendingSortByCreatedAtUtc.HasValue && !filter.IsAscendingSortByEntityName.HasValue)
+    {
+      changeLogs = filter.IsAscendingSortByCreatedAtUtc.Value
+        ? changeLogs.OrderBy(db => db.CreatedAtUtc)
+        : changeLogs.OrderByDescending(db => db.CreatedAtUtc);
+    }
+
+    if (filter.IsAscendingSortByEntityName.HasValue && !filter.IsAscendingSortByCreatedAtUtc.HasValue)
+    {
+      changeLogs = filter.IsAscendingSortByEntityName.Value
+              ? changeLogs.OrderBy(x => x.EntityName)
+              : changeLogs.OrderByDescending(x => x.EntityName);
+    }
+
+    if (filter.IsAscendingSortByEntityName.HasValue && filter.IsAscendingSortByCreatedAtUtc.HasValue)
+    {
+      changeLogs = (filter.IsAscendingSortByEntityName.Value
+        ? (filter.IsAscendingSortByCreatedAtUtc.Value
+          ? changeLogs.OrderBy(x => x.EntityName).ThenBy(x => x.CreatedAtUtc)
+          : changeLogs.OrderBy(x => x.EntityName).ThenByDescending(x => x.CreatedAtUtc))
+        : (filter.IsAscendingSortByCreatedAtUtc.Value
+          ? changeLogs.OrderByDescending(x => x.EntityName).ThenBy(x => x.CreatedAtUtc)
+          : changeLogs.OrderByDescending(x => x.EntityName).ThenByDescending(x => x.CreatedAtUtc)));
+    }
+
+    if (!string.IsNullOrWhiteSpace(filter.EntityNameIncludeSubstring))
+    {
+      changeLogs = changeLogs.Where(db => db.EntityName.Contains(filter.EntityNameIncludeSubstring));
+    }
+
+    if (!string.IsNullOrWhiteSpace(filter.PropertyNameIncludeSubstring))
+    {
+      changeLogs = changeLogs.Where(db => db.PropertyName.Contains(filter.PropertyNameIncludeSubstring));
+    }
+
+    return (
+      await changeLogs
+        .Skip(filter.SkipCount)
+        .Take(filter.TakeCount)
+        .ToListAsync(ct),
+      await changeLogs.CountAsync(ct));
   }
 
-  public Task<bool> RemoveAsync(Guid id, CancellationToken ct)
+  public async Task<bool> RemoveAsync(Guid id, CancellationToken ct)
   {
-    throw new NotImplementedException();
+    DbChangeLog changeLog = await _provider.ChangeLogs.FindAsync(id, ct);
+
+    if (changeLog is null)
+    {
+      return false;
+    }
+
+    _provider.ChangeLogs.Remove(changeLog);
+    await _provider.SaveAsync();
+
+    return true;
   }
```

Comment:
don't think this is required

Reply:
We won't delete the logs?

Reply:
I guess we will but not by single change log id but all of them by parent entity id

Reply:
++